### PR TITLE
Feature/aio 964 convert p2p onboarder to nfg UI

### PR DESCRIPTION
**Pull request by kylejmorin**
FP PR: https://github.com/network-for-good/Givecorps-site/pull/2337
NFG onboarder PR: https://github.com/network-for-good/nfg_onboarder/pull/88

### CODE REVIEW

**Comment by aravindpudayoor** [comment]
@hoenth Last day I had a discussion with @kylejmorin regarding this work. I think I need some more clarity as the above comment says that we have to use the high level for both single steps and group steps, but we have to render `/app/views/onboarding/_high_level_navigation.html.haml` and `/app/views/onboarding/_navigation.html.haml` files in separate classes too (the high level is inside the `.builder-nav` container and the navigation is inside the `.builder-container` container). Here I'm a bit uncertain about how to use both single steps, and group steps and use different presenters in the high-level file.

**Comment by hoenth** [comment]
@tmking Can you take a look at this with @aravindpudayoor

**Comment by tmking** [comment]
@aravindpudayoor if I'm understanding the situation correctly, shouldn't you be able to set two presenter objects in the controller, e.g. `group_nav_presenter` / `regular_nav_presenter`?

File: app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb [hunk 1]
```diff
@@ -20,25 +22,45 @@ def render_previous_button_unless?
     # for the step nav body text
     # or default to humanizing the
     # step's symbol.
-    def step_body(step)
-      I18n.t(step, scope: h.locale_namespace + [:step_navigations], default: step.to_s.humanize)
+    def step_body(nav_step)
+      I18n.t(
+        nav_step,
+        scope: h.locale_namespace + [:step_navigations],
+        default: nav_step.to_s.split('_').map(&:capitalize).join(' ')
```
Comment (hoenth) [inline app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb:29]: If you are trying to get each element capitalized, you can use `nav_step.to_s.humanize.titleize`

File: app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb [hunk 2]
```diff
@@ -20,25 +22,45 @@ def render_previous_button_unless?
...
-    def step_status(step)
-      return :active if step.to_sym == active_step.to_sym
-      return :visited if try(:completed_steps, h.controller_name).try(:include?, step)
-      # in case steps change, if the step or active step can't be found in what
-      # is currently the list of steps  don't barf
-      return :disabled if (all_steps.index(step.to_sym) || 0) > (all_steps.index(active_step.to_sym) || 0)
+    def step_status(nav_step)
+      return :active if active?(nav_step)
+      return :visited if visited?(nav_step)
+      return :disabled if disabled?(nav_step)
```
Comment (hoenth) [inline app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb:45]: is it expected that this returns nil if the step is not active, disabled or visited? Is there another state?

File: app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb [hunk 3]
```diff
@@ -20,25 +22,45 @@ def render_previous_button_unless?
...
+    end
+
+    def active?(nav_step)
+      nav_step.to_sym == h.controller_name.to_sym
```
Comment (hoenth) [inline app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb:49]: Why the change in the definition of active? Does `active_step` no longer contain the step which is active? If this is being used to display group steps, then the controller name won't match the step, as it is the controller is for the lower level steps.

File: app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb [hunk 4]
```diff
@@ -20,25 +22,45 @@ def render_previous_button_unless?
...
+    end
+
+    def disabled?(nav_step)
+      (all_steps.index(nav_step.to_sym) || 0) > (all_steps.index(active_step.to_sym) || 0)
+    end
```
Comment (hoenth) [inline app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb:54]: And above you use `active_step`. Seems if the definition of active step is changing, it should change in both places.

File: app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb [hunk 5]
```diff
@@ -20,25 +22,45 @@ def render_previous_button_unless?
...
+
+    def visited?(nav_step)
+      model.onboarder_progress.keys.include?(nav_step.to_s)
```
Comment (hoenth) [inline app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb:57]: I think the previous idea was that we need to keep track of each controller's progress. Does this still do that?

File: app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb [hunk 6]
```diff
@@ -20,25 +22,45 @@ def render_previous_button_unless?
...
+    end
+
+    def path_with_id(nav_step)
+      return nil unless model.onboarder_progress.key?(nav_step.to_s)
+
+      "/onboarding/create_campaign/#{nav_step}/#{model.onboarder_progress[nav_step.to_s].last}"
```
Comment (hoenth) [inline app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb:63]: I don't think we want to hard code this path here. The onboarder gem does not know anything about specific onboarders. It needs to work for all onboarders.

**Review by hoenth — COMMENTED** [review]
@aravindpudayoor I did not perform a complete review of this, as I perhaps don't understand what changes you are attempting to make. 

Perhaps we should have a discussion on what you are proposing here, so I can better understand them.

Is there a fundamental difference between how to two navigation bar presenters work? Don't they each list a set of steps, linking to previously visited steps, marking the current step as active, and disabling steps that have not yet been visited?

File: app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb (hunk 2 above)
Comment (aravindpudayoor) [reply app/presenters/nfg_onboarder/high_level_navigation_bar_presenter.rb:45]: I don't think there is another state other than these 3.

**Comment by kylejmorin** [comment]
@hoenth @tmking how do you feel about these changes? Do you feel like it is ready to merge in?

**Comment by kylejmorin** [comment]
@kcpavan could you review this work. Tom asked me to have you review it since he's given it some already. Thanks!

File: spec/dummy/db/migrate/20220603155503_add_save_and_exit_path_to_onboarding_session.rb [hunk 7]
```diff
@@ -0,0 +1,5 @@
+class AddSaveAndExitPathToOnboardingSession < ActiveRecord::Migration[5.2]
+  def change
+    add_column :onboarding_sessions, :save_and_exit_path, :string
```
Comment (kcpavan) [inline spec/dummy/db/migrate/20220603155503_add_save_and_exit_path_to_onboarding_session.rb:3]: Where are we using this?

**Review by kcpavan — COMMENTED** [review]
Overall looks fine, I have added a suggestion for you to consider.

File: spec/dummy/db/migrate/20220603155503_add_save_and_exit_path_to_onboarding_session.rb (hunk 7 above)
Comment (aravindpudayoor) [reply spec/dummy/db/migrate/20220603155503_add_save_and_exit_path_to_onboarding_session.rb]: @kcpavan This was used when we had some issue in the save & exit path. Since we are using the super method to save and exit we no longer need this now.

File: app/presenters/nfg_onboarder/wizard_steps_navigation_bar_presenter.rb [hunk 8]
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+module NfgOnboarder
+  # Ex usage on a view: NfgOnboarder::WizardStepsNavigationBarPresenter.new(onboarding_session, self)
+  class WizardStepsNavigationBarPresenter < NfgOnboarder::OnboarderPresenter
```
[the remark this answers is not in the document]
Comment (aravindpudayoor) [reply app/presenters/nfg_onboarder/wizard_steps_navigation_bar_presenter.rb]: Done!
